ocaml/xapi/xapi_vm_lifecycle.ml

```diff
@@ -279,7 +279,11 @@ let check_vgpu ~__context ~op ~ref_str ~vgpus ~power_state =
   | `pool_migrate | `migrate_send
     when List.for_all is_migratable  vgpus
       && List.for_all is_suspendable vgpus -> None
-  | `suspend | `checkpoint
+  | `checkpoint
+    when (List.for_all is_suspendable vgpus)
+      || (power_state = `Suspended
+      && List.for_all (fun vgpu -> Xapi_gpumon.Nvidia.is_nvidia ~__context ~vgpu) vgpus) -> None
```

Comment:
Is indentation correct here? The structure is `(a || (b && c))`.

Reply:
The logic looks correct to me.

Reply:
Yes, that's the structure.  `(a || (b && c))`

Reply:
Why do we need to check for nvidia again? I thought that the point of the helper functions was to hide that. Besides, if the power state is currently "suspended", then doesn't that already imply the right kind of vGPU?

Reply:
Emm, you are right. @robhoes  
If a VM with vGPU could went into a `suspended` state, `Nvidia` checking must already happened in `checkpoint`. 
I will refine it. Thanks.

Reply:
So the `is_suspendable` check is only valid for running VMs, not for suspended ones. So I think that we first need to check the power state. Therefore, I think that we should leave the existing case as is, and add the following above it:
```
| `checkpoint
  when power_state = `Suspended -> None
```